lib/kube/proxy/server.go

```diff
@@ -261,9 +261,23 @@ func (t *TLSServer) Serve(listener net.Listener) error {
 
 // Close closes the server and cleans up all resources.
 func (t *TLSServer) Close() error {
+	return trace.Wrap(t.close(t.closeContext))
+}
+
+// Close closes the server and cleans up all resources.
+func (t *TLSServer) Shutdown(ctx context.Context) error {
+	// TODO(tigrato): handle connections gracefully and wait for them to finish.
+	// This might be problematic because exec and port forwarding connections
+	// are long lived connections and if we wait for them to finish, we might
+	// end up waiting forever.
+	return trace.Wrap(t.close(ctx))
+}
+
+// close closes the server and cleans up all resources.
+func (t *TLSServer) close(ctx context.Context) error {
 	var errs []error
 	for _, kubeCluster := range t.fwd.kubeClusters() {
-		errs = append(errs, t.unregisterKubeCluster(t.closeContext, kubeCluster.GetName()))
+		errs = append(errs, t.unregisterKubeCluster(ctx, kubeCluster.GetName()))
 	}
 	errs = append(errs, t.fwd.Close(), t.Server.Close())
```

Comment:
> This might be problematic because exec and port forwarding connections
> are long lived connections and if we wait for them to finish, we might
> end up waiting forever.

At least the in-process reload passes in a timeout context with the shutdown timeout. not sure why we don't have a timeout on shutdown by SIGQUIT and SIGUP.

Reply:
Right now I am closing the connections without sending the error to the user. Kube protocol requires sending the error messages to a special stream or "channel" depending on the underlying protocol they use in order to properly inform the client of what happened.